pkg/planner/cascades/memo/memo.go

```diff
@@ -16,20 +16,36 @@ package memo
 
 import (
 	"container/list"
+	"sync"
 
 	base2 "github.com/pingcap/tidb/pkg/planner/cascades/base"
 	"github.com/pingcap/tidb/pkg/planner/core/base"
-	"github.com/pingcap/tidb/pkg/sessionctx"
 	"github.com/pingcap/tidb/pkg/util/intest"
 )
 
+// MemoPool is initialized for memory saving by reusing TaskStack.
+var MemoPool = sync.Pool{
+	New: func() any {
+		return NewMemo()
+	},
+}
+
+// Destroy indicates that when stack itself is useless like in the end of optimizing phase, we can destroy ourselves.
+func (mm *Memo) Destroy() {
+	// when a TaskStack itself is useless, we can destroy itself actively.
+	mm.groupIDGen.id = 0
+	mm.rootGroup = nil
+	mm.groups.Init()
+	clear(mm.groupID2Group)
+	clear(mm.hash2GroupExpr)
+	mm.hasher.Reset()
+	MemoPool.Put(mm)
+}
```

Comment:
Is necessary to create a pool for the `Memo` whose lifetime is not very short?
/cc @hawkingrei

Reply:
Early optimization is a taboo in software design.